packages/scripts/src/clusterize.ts

```diff
@@ -0,0 +1,71 @@
+import dbscan from "@cdxoo/dbscan";
+import { strict as assert } from "assert";
+
+/**
+  Based on https://www.npmjs.com/package/@cdxoo/dbscan
+ */
+export type DbscanOptions = {
+  /**
+    Maximum distance between datapoints. Determine if a datapoint is in a
+    cluster or not. Default is 1.0.
+   */
+  epsilon: number;
+
+  /**
+    Custom function to compare calculated distance and epsilon. Default is
+    (dist, e) => (dist < e)
+   */
+  epsilonCompare: (distance: number, epsilon: number) => boolean;
+
+  /**
+    Threshold of how many points are needed in the same neighborhood to form a
+    cluster. Default is 2.
+   */
+  minimumPoints: number;
+};
+
+const defaultDbscanOptions: DbscanOptions = {
+  epsilon: 1.0,
+  epsilonCompare: (d, e) => d < e,
+  minimumPoints: 2,
+};
+
+export const clusterize = <T>(
+  dataset: T[],
+  getVector: (x: T) => number[],
+  optionsIn?: Partial<DbscanOptions>
+): { clusters: T[][]; noise: T[] } => {
+  if (dataset.length === 0) {
+    return { clusters: [], noise: [] };
+  }
+
+  const options: DbscanOptions = {
+    ...defaultDbscanOptions,
+    ...(optionsIn ?? {}),
+  };
+
+  const { clusters: clusterIndexesArray, noise: noiseIndexes } = dbscan({
+    ...options,
+    dataset,
+    distanceFunction(a, b) {
+      return distanceSquared(getVector(a), getVector(b));
+    },
+  });
+
+  return {
+    clusters: clusterIndexesArray.map((indexes) =>
+      indexes.map((index) => dataset[index])
+    ),
+    noise: noiseIndexes.map((index) => dataset[index]),
+  };
+};
+
+export const distanceSquared = (a: number[], b: number[]) => {
+  assert(
+    a.length === b.length,
+    "Unexpected inconsistent dimensionality in distanceFunction"
+  );
+  // Pythagorean theorem: the square of the hypotenuse is equal to the sum of
+  // the squares of the other sides
+  return a.reduce((acc, cur, i) => acc + Math.pow(b[i] - cur, 2), 0);
```

Comment:
ditto this comment https://github.com/mongodb/chatbot/pull/292/files#r1447931008

Reply:
and again ChatGPT generated answer w/ tensorflow:

```ts
export const distanceSquared = async (a: number[], b: number[]) => {
  // Convert arrays to tensors if they are not already
  const tensorA = tf.tensor(a);
  const tensorB = tf.tensor(b);

  // Check dimensions
  if (tensorA.shape[0] !== tensorB.shape[0]) {
    throw new Error("Unexpected inconsistent dimensionality in distanceFunction");
  }

  // Calculate squared distance
  const squaredDistanceTensor = tensorA.sub(tensorB).square().sum();

  // Get the result as a number
  const squaredDistance = await squaredDistanceTensor.array();
  
  // Since the result is a scalar, we can return the first element of the array
  return squaredDistance[0];
};
```

Reply:
Pretty cool, I wish I had thought of tensorflow before